core/trino-main/src/main/java/io/trino/operator/scalar/ConcatFunction.java

```diff
@@ -20,43 +20,85 @@
 
 import java.lang.invoke.MethodHandle;
+import java.util.List;
+import java.util.Optional;
 
+import static com.google.common.collect.Iterables.getOnlyElement;
 import static io.trino.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
 import static io.trino.spi.StandardErrorCode.NOT_SUPPORTED;
 import static io.trino.spi.block.PageBuilderStatus.DEFAULT_MAX_PAGE_SIZE_IN_BYTES;
 import static io.trino.spi.function.InvocationConvention.InvocationArgumentConvention.NEVER_NULL;
 import static io.trino.spi.function.InvocationConvention.InvocationReturnConvention.FAIL_ON_NULL;
+import static io.trino.spi.type.TypeSignatureParameter.typeVariable;
 import static io.trino.spi.type.VarbinaryType.VARBINARY;
+import static io.trino.spi.type.VarcharType.MAX_LENGTH;
 import static io.trino.spi.type.VarcharType.VARCHAR;
+import static io.trino.spi.type.VarcharType.createVarcharType;
 import static io.trino.util.Reflection.methodHandle;
 import static java.lang.Math.addExact;
+import static java.lang.Math.toIntExact;
 import static java.util.Collections.nCopies;
 
 public final class ConcatFunction
         extends SqlScalarFunction
 {
-    // TODO design new variadic functions binding mechanism that will allow to produce VARCHAR(x) where x < MAX_LENGTH.
-    public static final ConcatFunction VARCHAR_CONCAT = new ConcatFunction(VARCHAR.getTypeSignature(), "Concatenates given strings");
+    public static final ConcatFunction VARCHAR_CONCAT = new ConcatFunction(VARCHAR);
 
-    public static final ConcatFunction VARBINARY_CONCAT = new ConcatFunction(VARBINARY.getTypeSignature(), "concatenates given varbinary values");
+    public static final ConcatFunction VARBINARY_CONCAT = new ConcatFunction(VARBINARY);
 
     private static final int MAX_INPUT_VALUES = 254;
     private static final int MAX_OUTPUT_LENGTH = DEFAULT_MAX_PAGE_SIZE_IN_BYTES;
 
-    private ConcatFunction(TypeSignature type, String description)
+    private ConcatFunction(Type type)
     {
-        super(FunctionMetadata.scalarBuilder()
-                .signature(Signature.builder()
-                        .name("concat")
-                        .returnType(type)
-                        .argumentType(type)
-                        .variableArity()
-                        .build())
-                .description(description)
-                .build());
+        super(buildFunctionMetadata(type));
+    }
+
+    private static FunctionMetadata buildFunctionMetadata(Type type)
+    {
+        if (type == VARCHAR) {
+            return FunctionMetadata.scalarBuilder()
+                    .signature(Signature.builder()
+                            .name("concat")
+                            .returnTypeDerivation(Optional.of(ConcatFunction::calculateReturnType))
+
+                            // still required for SHOW FUNCTIONS result
+                            .longVariable("x", "1")
+                            .returnType(new TypeSignature("varchar", typeVariable("x")))
```

Comment:
Can we use unbounded varchar as a "fake" return type?

Reply:
Probably, it requires additional change. Below exception happens after changing to `returnType(VARCHAR)` without `longVariable`.
```
Bound signature concat(varchar, varchar):varchar(3) does not match declared signature concat(varchar):varchar
java.lang.IllegalArgumentException: Bound signature concat(varchar, varchar):varchar(3) does not match declared signature concat(varchar):varchar
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:435)
	at io.trino.metadata.SignatureBinder.verifyBoundSignature(SignatureBinder.java:332)
	at io.trino.metadata.SignatureBinder.extractBoundVariables(SignatureBinder.java:306)
	at io.trino.metadata.SignatureBinder.bindFunction(SignatureBinder.java:251)
	at io.trino.metadata.FunctionResolver.toFunctionBinding(FunctionResolver.java:127)
	at io.trino.metadata.FunctionResolver.matchFunction(FunctionResolver.java:258)
	at io.trino.metadata.FunctionResolver.matchFunctionWithCoercion(FunctionResolver.java:233)
	at io.trino.metadata.FunctionResolver.resolveFunction(FunctionResolver.java:174)
	at io.trino.metadata.MetadataManager.resolvedFunctionInternal(MetadataManager.java:2115)
	at io.trino.metadata.MetadataManager.lambda$resolvedFunctionInternal$49(MetadataManager.java:2110)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at io.trino.metadata.MetadataManager.resolvedFunctionInternal(MetadataManager.java:2110)
	at io.trino.metadata.MetadataManager.resolveFunction(MetadataManager.java:2081)
	at io.trino.sql.analyzer.ExpressionAnalyzer$Visitor.visitFunctionCall(ExpressionAnalyzer.java:1258)
	at io.trino.sql.analyzer.ExpressionAnalyzer$Visitor.visitFunctionCall(ExpressionAnalyzer.java:581)
	at io.trino.sql.tree.FunctionCall.accept(FunctionCall.java:121)
	at io.trino.sql.tree.StackableAstVisitor.process(StackableAstVisitor.java:27)
	at io.trino.sql.analyzer.ExpressionAnalyzer$Visitor.process(ExpressionAnalyzer.java:604)
	at io.trino.sql.analyzer.ExpressionAnalyzer.analyze(ExpressionAnalyzer.java:485)
	at io.trino.sql.analyzer.ExpressionAnalyzer.analyzeExpression(ExpressionAnalyzer.java:3441)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.analyzeExpression(StatementAnalyzer.java:4075)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.analyzeSelectSingleColumn(StatementAnalyzer.java:3887)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.analyzeSelect(StatementAnalyzer.java:3673)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.visitQuerySpecification(StatementAnalyzer.java:2436)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.visitQuerySpecification(StatementAnalyzer.java:457)
	at io.trino.sql.tree.QuerySpecification.accept(QuerySpecification.java:155)
	at io.trino.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:474)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:482)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:1394)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:457)
	at io.trino.sql.tree.Query.accept(Query.java:107)
	at io.trino.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at io.trino.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:474)
	at io.trino.sql.analyzer.StatementAnalyzer.analyze(StatementAnalyzer.java:436)
	at io.trino.sql.analyzer.Analyzer.analyze(Analyzer.java:79)
	at io.trino.sql.analyzer.Analyzer.analyze(Analyzer.java:71)
	at io.trino.execution.SqlQueryExecution.analyze(SqlQueryExecution.java:261)
	at io.trino.execution.SqlQueryExecution.<init>(SqlQueryExecution.java:199)
	at io.trino.execution.SqlQueryExecution$SqlQueryExecutionFactory.createQueryExecution(SqlQueryExecution.java:825)
	at io.trino.dispatcher.LocalDispatchQueryFactory.lambda$createDispatchQuery$0(LocalDispatchQueryFactory.java:135)
	at io.trino.$gen.Trino_testversion____20220914_081840_71.call(Unknown Source)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:131)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:74)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:82)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

Reply:
It doesn't make sense to allow a function definition to declare a return type and a returnTypeDerivation at the same time:
* If they are not the same value, which one matters? 
* If they are the same value, why have the returnTypeDerivation at all?

Reply:
What should be the return type printed by SHOW FUNCTIONS?